File: packages/provide-uterm-go/session/session.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

// Package session defines the terminal-session interfaces shared by
// transports, emulation, and I/O helpers, plus ports of
// provide.uterm.io (PromptWaiter, InputSender) and provide.uterm.expect
// (SendAndExpect).
package session

import (
	"context"
	"time"
)

// Cursor is a screen cursor position.
type Cursor struct {
	X int `json:"x"`
	Y int `json:"y"`
}

// PromptDetection is the prompt-detector result attached to a snapshot.
// Fields mirror the Python detection dict keys.
type PromptDetection struct {
	PromptID  string         `json:"prompt_id"`
	InputType string         `json:"input_type,omitempty"`
	IsIdle    bool           `json:"is_idle"`
	KVData    map[string]any `json:"kv_data,omitempty"`
	Extra     map[string]any `json:"-"`
}

// Snapshot is the emulated screen state. It mirrors the Python
// TerminalEmulator.get_snapshot() dict.
type Snapshot struct {
	Screen           string           `json:"screen"`
	ScreenHash       string           `json:"screen_hash"`
	Cursor           Cursor           `json:"cursor"`
	Cols             int              `json:"cols"`
	Rows             int              `json:"rows"`
	Term             string           `json:"term"`
	CursorAtEnd      bool             `json:"cursor_at_end"`
	HasTrailingSpace bool             `json:"has_trailing_space"`
	RawTail          string           `json:"raw_tail"`
	CapturedAt       float64          `json:"captured_at"`
	PromptDetected   *PromptDetection `json:"prompt_detected,omitempty"`
}

// Session is the minimal interface expected by PromptWaiter and InputSender
// (the Python io.Session protocol).
type Session interface {
	// WaitForUpdate blocks until new bytes arrive from the remote or the
	// timeout elapses; it reports whether new data arrived.
	WaitForUpdate(ctx context.Context, timeout time.Duration) (bool, error)
	// Snapshot returns the latest screen state without performing network I/O.
	Snapshot() Snapshot
	// Send writes data to the session.
	Send(ctx context.Context, data string) error
}

// ConnectionChecker is optionally implemented by sessions that can report
// their connection state (the Python code duck-types is_connected).
type ConnectionChecker interface {
	IsConnected() bool
}

// IdleReporter is optionally implemented by sessions that can predict when
// the screen will be considered idle (the Python seconds_until_idle hook).
type IdleReporter interface {
	SecondsUntilIdle() float64
}

// ExpectSession is the interface required by SendAndExpect (the Python
// expect.ExpectSession protocol).
type ExpectSession interface {
	Send(ctx context.Context, data string) error
	Snapshot() Snapshot
	ScreenChangeSeq() int
	// WaitForScreenChange blocks until the screen updates beyond since (or
	// any next update when since < 0), reporting whether it changed.
	WaitForScreenChange(ctx context.Context, timeout time.Duration, since int) (bool, error)
}

// sessionConnected mirrors Python _session_is_connected: sessions without an
// is_connected implementation are treated as connected.
func sessionConnected(s any) bool {
	checker, ok := s.(ConnectionChecker)
	if !ok {
		return true
	}
	return checker.IsConnected()
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/Session/Session.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

namespace Provide.Uterm.Session;

public sealed class Cursor
{
    public int X { get; set; }
    public int Y { get; set; }
}

/// <summary>Struct form of cursor position used by some call sites.</summary>
public struct CursorPos
{
    public int X { get; set; }
    public int Y { get; set; }

    public CursorPos(int x, int y)
    {
        X = x;
        Y = y;
    }
}

public sealed class PromptDetection
{
    public string PromptId { get; set; } = "";
    public string InputType { get; set; } = "";
    public bool IsIdle { get; set; }
    public Dictionary<string, object?> KvData { get; set; } = new();
    public Dictionary<string, object?> Extra { get; set; } = new();
}

/// <summary>Emulated screen state. Mirrors TerminalEmulator.GetSnapshot().</summary>
public sealed class Snapshot
{
    public string Screen { get; set; } = "";
    public string ScreenHash { get; set; } = "";
    public Cursor Cursor { get; set; } = new();
    public int Cols { get; set; }
    public int Rows { get; set; }
    public string Term { get; set; } = "ANSI";
    public bool CursorAtEnd { get; set; }
    public bool HasTrailingSpace { get; set; }
    public string RawTail { get; set; } = "";
    public double CapturedAt { get; set; }
    public PromptDetection? PromptDetected { get; set; }
}

/// <summary>Minimal session protocol for PromptWaiter / InputSender.</summary>
public interface ISession
{
    Task<bool> WaitForUpdateAsync(TimeSpan timeout, CancellationToken cancellationToken = default);
    Snapshot Snapshot();
    Task SendAsync(string data, CancellationToken cancellationToken = default);
}

public interface IConnectionChecker
{
    bool IsConnected();
}

public interface IExpectSession
{
    Task SendAsync(string data, CancellationToken cancellationToken = default);
    Snapshot Snapshot();
    int ScreenChangeSeq();
    Task<bool> WaitForScreenChangeAsync(TimeSpan timeout, int since, CancellationToken cancellationToken = default);
}

public sealed class ExpectOptions
{
    public string? ExpectText { get; set; }
    public string? ExpectRegex { get; set; }
    public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(10);
}

public sealed class ExpectResult
{
    public bool Matched { get; set; }
    public Snapshot Snapshot { get; set; } = new();
    public string Screen { get; set; } = "";
}

/// <summary>Send keys and wait for expected terminal output.</summary>
public static class Expect
{
    public static async Task<ExpectResult> SendAndExpectAsync(
        IExpectSession session,
        string keys,
        ExpectOptions? options = null,
        CancellationToken cancellationToken = default)
    {
        options ??= new ExpectOptions();
        var since = session.ScreenChangeSeq();
        await session.SendAsync(keys, cancellationToken);
        var deadline = DateTime.UtcNow + options.Timeout;
        while (DateTime.UtcNow < deadline)
        {
            var remaining = deadline - DateTime.UtcNow;
            if (remaining <= TimeSpan.Zero)
            {
                break;
            }

            await session.WaitForScreenChangeAsync(remaining, since, cancellationToken);
            var snap = session.Snapshot();
            var screen = snap.Screen;
            var matched = false;
            if (!string.IsNullOrEmpty(options.ExpectText) && screen.Contains(options.ExpectText, StringComparison.Ordinal))
            {
                matched = true;
            }
            else if (!string.IsNullOrEmpty(options.ExpectRegex) &&
                     System.Text.RegularExpressions.Regex.IsMatch(screen, options.ExpectRegex))
            {
                matched = true;
            }
            else if (string.IsNullOrEmpty(options.ExpectText) && string.IsNullOrEmpty(options.ExpectRegex))
            {
                matched = true;
            }

            if (matched)
            {
                return new ExpectResult { Matched = true, Snapshot = snap, Screen = screen };
            }

            since = session.ScreenChangeSeq();
        }

        var final = session.Snapshot();
        return new ExpectResult { Matched = false, Snapshot = final, Screen = final.Screen };
    }
}
